noise/README.md

```diff
@@ -0,0 +1,400 @@
+section](#design-considerations-and-tradeoffs) covers the rationale behind the
+presented design.
+
+libp2p peers have an "identity keypair" which is used to derive their peer id,
+as described in the [peer id spec][peer-id-spec]. noise-libp2p uses the identity
+keypair to authenticate a separate static key which is used only for
+noise-libp2p. This is described in the [Static Key Authentication
+section](#static-key-authentication).
+
+## Message Framing
+
+noise-libp2p sends all data over the wire in the form of Noise messages, which
+are [defined][npf-message-format] with a maximum length of 65535 bytes.
+
+Noise messages are sent prefixed with their length in bytes, encoded as an
+unsigned 16-bit integer in network (big-endian) byte order.
+
+## Supported DH Functions, Ciphers, and Hash Functions
+
+The Noise framework allows protocol designers to choose from a small set of
+Diffie-Hellman functions, symmetric ciphers, and hash functions.
+
+For simplicity, noise-libp2p currently supports one function from each category:
+
+### Supported DH Functions
+
+noise-libp2p supports the 25519 DH functions [as defined in the Noise
+spec][npf-dh-25519].
+
+### Supported Cipher Functions
+
+noise-libp2p supports the ChaChaPoly cipher functions [as defined in the Noise
+spec][npf-cipher-chachapoly].
+
+### Supported Hash Functions
+
+noise-libp2p supports the SHA256 hash function [as defined in
+the Noise spec][npf-hash-sha256].
+
+## Supported Handshake Patterns
+
+Noise defines twelve [fundamental interactive handshake
+patterns][npf-fundamental-patterns] for exchanging public keys between parties and performing
+Diffie-Hellman computations.
+
+The patterns are named according to whether static keypairs are used, and if so,
+by what means each party gains knowledge of the other's static public key.
+
+From the [fundamental patterns section of the Noise spec][npf-fundamental-patterns]:
+
+    The first character refers to the initiator's static key:
+
+       * N = No static key for initiator
+       * K = Static key for initiator Known to responder
+       * X = Static key for initiator Xmitted ("transmitted") to responder
+       * I = Static key for initiator Immediately transmitted to responder, despite reduced or absent identity hiding
+
+    The second character refers to the responder's static key:
+
+      * N = No static key for responder
+      * K = Static key for responder Known to initiator
+      * X = Static key for responder Xmitted ("transmitted") to initiator
+
+noise-libp2p supports a [compound protocol][npf-compound-protocols] consisting
+of two fundamental handshake patterns, `XX` and `IK`. A variation of the `XX`
+pattern, `XXfallback` is used to gracefully recover from a failed `IK`
+handshake attempt.
+
+This pattern of using `XX`, `IK` and `XXfallback` is described in the Noise spec
+as the [Noise Pipes compound protocol][npf-noise-pipes].
+
+The`XX` pattern is used when the initiator (or in libp2p terms, the dialing peer)
```

Comment:
Yeah, I think you're right that we should encourage static keys to have a long lifetime. How about implementations MUST support saving and loading static Noise keys, and we add a recommendation to only change the Noise key if you suspect it's been compromised?

Reply:
mandating it seems overly restrictive and unenforceable (on the wire). a strong recommendation? since the advantage is obvious it seems that clients have little incentive to avoid storing on a whim - but those with a legit reason (tailored for read-only devices for example) get locked out.

Reply:
A strong SHOULD with an overview of the implications of not complying would work for me.

Reply:
I think the choice of lifetime and storage of static DH keys should generally be left to the implementation and ultimately the concrete application. Given that in the context of libp2p a peer already has to (securely) store the identity keypair, I don't think that additionally persisting the DH static keypair somewhere is strictly advantageous, considering that it increases exposure of that keypair. In this context, only keeping static DH keypairs around in-memory, e.g. for the lifetime of a process, seems quite attractive. I am thus in favor of a wording that states that support and use of the Noise Pipes compound protocol (and support for `XXfallback`) is strictly optional but may be recommended for increased efficiency. If necessary for protocol negotiation, support for `XXfallback` could be bound to support for `IK`, so that an implementation must either support both of these or none, leaving support for `XX` and possibly `IX` as mandatory.

Reply:
@romanb makes a good point that persisting the keys comes with increased exposure. BTW, I wasn't saying that persisted keys MUST be used, just that the API should allow you to use persisted keys... but even that seems a bit too prescriptive.

I think you're right that making Noise Pipes optional is a good call; it's a fair bit of complexity for something that not everyone will want. If you're fine with `IX`, for example, there's no efficiency benefit to using Noise Pipes, since you get 1-RTT every time with IX.

I'd be fine with removing Noise Pipes (and `XXfallback`) from the spec for this iteration, actually. It would simplify things quite a bit, and we could mention in the design consideration section that it's a possible future enhancement. For now I'll put it into its own section and describe it as an optional optimization.